File: code/matlab/dyn_example1.m
clear all, close all, clc

A = [1 2; 4 3]
T = [-1 0.5; 1 1]
D = [-1 0; 0 5]

N = 10;
T = 1.0;
t = linspace(0, T, N+1);
x0 = [1; 1]
x = zeros(2, N+1)
for i = 1:N+1
    x(:, i) = T*exp(D*t(i))*inv(T)*x0;
end
x

hold on
plot(t,x(1,:), 'o-')
plot(t,x(2,:), 'o-')
legend(["x_1" "x_2"])
xlabel('t')
ylabel('x')
grid()
hold off

File: code/python/dyn_example1.py
import numpy as np
from scipy.linalg import expm
import matplotlib.pyplot as plt

A = np.array([[1, 2], [4, 3]])
D, T = np.linalg.eig(A)
# These work also:
#T = [-1 0.5; 1 1];
#D = [-1 0; 0 5];

n = 8
t_n = 0.4
x0 = np.array([[1], [1]])
T_inv = np.linalg.inv(T)
t_range = np.linspace(0, t_n, n+1)
x = [T.dot(expm(np.diag(D)*t)).dot(T_inv).dot(x0) for t in t_range]
x = np.hstack(x)

plt.plot(t_range, x[0], 'o-', label='$x_1$')
plt.plot(t_range, x[1], 'o-', label='$x_2$')
plt.legend()
plt.xlabel('t')
plt.ylabel('x')
plt.grid()
plt.show()
